feat: basic Support for OpenTelemetry Metrics and Token Usage Metrics in OpenAI V1

Humbertzhang opened the pull request:
Hello all, this is my first PR for [issue#251](https://github.com/traceloop/openllmetry/issues/251) 

This PR introduces fundamental components for OpenTelemetry Metrics support, and it add counters for token usage data of `openai.resources.chat.completions`  in OpenAI V1.

My goal with this PR is to present my approach to addressing the issue at hand and to seek your feedback on the implementation. Your insights and suggestions will be extremely valuable for refining this solution. Moving forward, based on your feedback, I plan to enhance this implementation further and extend support to other instrumentors.

Looking for your reviews and constructive criticism to help improve this integration.

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/traceloop/openllmetry?pullRequest=369) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/traceloop/openllmetry?pullRequest=369) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/traceloop/openllmetry?pullRequest=369) it.</sub>

nirga commented:
Nice work @Humbertzhang! Let's connect it to some observability platform to see if it works?

nirga commented:
Sharing some work that's being done now with the OpenTelemetry community that we should align with - https://github.com/traceloop/semantic-conventions/pull/2

Humbertzhang commented:
Hi @nirga!
I setup a demo environment using `otel-collector` and `prometheus`, and then setup a openai + traceloop demo for generate metrics.
You can find those files at: https://github.com/Humbertzhang/demo_otel_prometheus

And I think it reports metrics as expect!

<img width="1728" alt="image" src="https://github.com/traceloop/openllmetry/assets/24400940/f980d034-4bee-425b-9e23-6e5e5057b213">

Humbertzhang commented:
> Sharing some work that's being done now with the OpenTelemetry community that we should align with - [traceloop/semantic-conventions#2](https://github.com/traceloop/semantic-conventions/pull/2)

ok @nirga , I will look into this pr  align with it !

Humbertzhang commented:
#### Update to Align with https://github.com/traceloop/semantic-conventions/pull/2
Hello @nirga, I have updated my code to align with the changes proposed of `openai.chat_completions.tokens` metric in https://github.com/traceloop/semantic-conventions/pull/2.

You can see the results under normal conditions in the attached image. I have appropriately added attributes such as `llm.response.model` and `llm.usage.token_type` .

![20240131-210801](https://github.com/traceloop/openllmetry/assets/24400940/9a8a51a5-e47d-4833-80a3-550feb8ad8dc)

You can observe the results under exceptional conditions in below images.
During this run, the network was initially stable but became unreachable for OpenAI requests after a few minutes, resulting in an APIConnectionError.

![20240131-210757](https://github.com/traceloop/openllmetry/assets/24400940/4f0ec57b-e407-43f4-acce-b8123458fbbf)
![20240131-210751](https://github.com/traceloop/openllmetry/assets/24400940/876a2339-b61c-4deb-90ee-dc7aeccca30f)

However, I encountered challenges in retrieving the `server.address`. I attempted to acquire it in the same manner as the Span(Like https://github.com/traceloop/openllmetry/blob/main/packages/opentelemetry-instrumentation-openai/opentelemetry/instrumentation/openai/shared/__init__.py#L40), but this it did not get the base_url in practical scenarios(). Do you have any alternative suggestions or insights on this matter?

nirga commented:
Looks good @Humbertzhang! Before I review it, can you:
* rebase
* fix lint issues
* add tests (this actually helped me better review PRs in the past 😅)

Reg. your question - can you give an example of when it didn't work for you?

Humbertzhang commented:
Hi @nirga !
I have added all the suggested metrics of openai and their corresponding tests in this pr:
* `llm.openai.chat_completions.tokens`
* `llm.openai.chat_completions.choices`
* `llm.openai.chat_completions.duration`
* `llm.openai.embeddings.tokens`
* `llm.openai.embeddings.vector_size`
* `llm.openai.embeddings.duration`
* `llm.openai.image_generations.duration`

and I have manually tested all metrics in prometheus.

Additionally, I've completed the rebase and addressed all linting issues.

Regarding `server.address`, I add a function called `_get_openai_base_url`, and I always get a `""` response by calling it. 
This function is intended to mirror the method used by spans to retrieve the OpenAI URL. 
I'm uncertain if there's a more effective approach to obtain the OpenAI URL—any suggestions would be appreciated.

For the metric `Metric: openai.embeddings.vector_size`, I'm contemplating whether a counter is the most suitable instrument for recording it. Maybe [Gauge](https://opentelemetry.io/docs/specs/otel/metrics/api/#gauge) is better for recording it? What do you think about it?

I look forward to your review and any feedback you may have!

Humbertzhang commented:
Hi @nirga , I have updated tests for metrics using VCR.

gyliu513 commented:
Thanks @Humbertzhang !

Hey @nirga , can we get this merged? I was planning to create a PR to enable watsonx as well, and it will depend on this PR, thanks!

nirga commented:
@gyliu513 yeah probably today / tomorrow. I need to see why the tests are failing, and I want to move this to a common semantic conventions.

nirga commented:
@Humbertzhang looks like there's regression in some tests? 🤔

Humbertzhang commented:
Hi @nirga, there's a bug that causing the failure of streaming tests, and i have fixed it, u can try to run tests again!

nirga commented:
Looks like it's still failing @Humbertzhang. Does it work for you locally?

Humbertzhang commented:
The previous streaming releated failure is fixed, but there are still some other failure that i am trying to fix. I add metrics tests in `traceloop-sdk/tests/metrics` , and it breaks other tests in `traceloop-sdk/tests/`. I'm looking at what causes them to interact with each other (which wasn't a few days ago)

nirga reviewed (commented):
Overall this looks good! @Humbertzhang I fixed the tests and moved them to be under OpenAI (it makes more sense IMO since these are specific to OpenAI and not the SDK).

I also left a couple of comments, and note that you need to fix some linting issues you have. 

One open question I still have is whether the separation between the metrics accumulation and trace building makes sense - it requires some amount of code duplication where we're basically taking the same attributes from the request and the response just to put them in a trace or in metrics. Have you considered uniting them? What are the pros / cons?

Humbertzhang commented:
Thank you for your insights and for integrating the tests under OpenAI, which indeed makes more sense.

On the topic of separating metrics accumulation from trace building, my initial intent was to minimize the need for altering existing functions like `_with_tracer_wrapper` and `chat_wrapper`, aiming to preserve the current codebase integrity. This approach was chosen to avoid significant disruptions. However, I acknowledge the duplication issue you've pointed out, where we're handling similar attributes in both metrics and traces.

While I had concerns about the complexity and potential bugs that merging these functionalities might introduce, I see the merit in reconsidering this strategy to reduce redundancy. I'm open to further discussions to refine our approach, balancing minimal code impact with efficiency.

nirga commented:
Let's do that then @Humbertzhang. I think it will result in cleaner code

Humbertzhang commented on packages/opentelemetry-instrumentation-openai/opentelemetry/instrumentation/openai/shared/chat_wrappers.py:
```diff
@@ -27,6 +30,73 @@
+            "server.address": _get_openai_base_url(),
+        }
+
+        token_counter.add(1, attributes=attributes)
```
The reason for incrementing by 1 in the case of an exception is to reflect the occurrence of errors in our metrics, given that we might not be able to obtain the actual token_count and choice_count due to the exception. I thought that incrementing by 1 could serve as a simple indicator of error occurrences in our metrics tracking. Do you think this approach needs adjustment? If so, how would you suggest we modify it to better capture our needs?

nirga replied:
Let's not capture the errors in the metric maybe, just in the trace

nirga replied:
(potentially we can have another metric to count errors or something)

Humbertzhang commented on packages/opentelemetry-instrumentation-openai/opentelemetry/instrumentation/openai/shared/chat_wrappers.py:
```diff
@@ -27,6 +30,73 @@
+        start_time = time.time()
+        response = wrapped(*args, **kwargs)
+        end_time = time.time()
+    except Exception as e:  # pylint: disable=broad-except
```
I added these operations on exception following the semantic conventions, which suggest including `error.type` for metrics if the operation ends in error. Do you think we still need to handle `error.type` in this context?

nirga replied:
hmm the only thing that might be weird here is that we're increasing the token count, no?

Humbertzhang replied:
Yes, it's quite a little odd😂...Maybe we should skip record `error.type` in metrics?

Humbertzhang commented on packages/opentelemetry-instrumentation-openai/opentelemetry/instrumentation/openai/v1/__init__.py:
```diff
@@ -39,6 +47,57 @@ def _instrument(self, **kwargs):
             completion_wrapper(tracer),
         )
+
+        chat_token_counter = meter.create_counter(
```
No, it doesn't throw exceptions when metering is disabled, but leads to unnecessary metrics records; adding checks before all meter.create_xxx to avoid this.

Humbertzhang commented on packages/opentelemetry-instrumentation-openai/opentelemetry/instrumentation/openai/v1/__init__.py:
```diff
@@ -2,20 +2,24 @@
 from opentelemetry.instrumentation.openai.shared.embeddings_wrappers import (
     embeddings_wrapper,
-    aembeddings_wrapper,
+    aembeddings_wrapper, embeddings_metrics_wrapper,
```
Manually formatted here

Humbertzhang commented:
Hi @nirga, I've integrated the metrics recording with the previous processes, ran tests locally, and conducted manual testing with Prometheus. Looking forward to your feedback!

Humbertzhang commented:
Resolved poetry.lock issue affecting tests. Please re-run tests for it.

nirga commented:
Nice work @Humbertzhang! It's a significant milestone for OpenLLMetry :)

paolorechia commented:
@Humbertzhang thanks for the work here, could you resolve the conflicts so we can merge? :)

paolorechia commented on packages/opentelemetry-instrumentation-openai/opentelemetry/instrumentation/openai/shared/__init__.py:
```diff
@@ -129,6 +131,13 @@ def _set_response_attributes(span, response):
         )
 
 
+def _get_openai_base_url():
```
:question: Is this function still always returning empty string?

I can confirm this is behavior if no initialization is done.
```python
In [2]: def _get_openai_base_url():
   ...:     base_url = openai.base_url if hasattr(openai, "base_url") else openai.api_base
   ...:     if not base_url:
   ...:         return ""
   ...:     return base_url
   ...: 

In [3]: import openai

In [4]: _get_openai_base_url()
Out[4]: ''

In [5]: 
```

On the other hand, if you inspect the client instance, you should get the base url:

```python
In [6]: client = openai.OpenAI()

In [7]: client.base_url
Out[7]: URL('https://api.openai.com/v1/')
```

Does this help you?

paolorechia replied:
I've reviewed your code, sadly I don't see an easy way to extract this URL from an instance in the current instrumentation code. 

My only idea was to instrument the client constructor so you can inject a listener which retrieves the URL for you. If you store it in a Singleton, you could retrieve it while calling this function:

```
def _handle_response(response, span, token_counter=None, choice_counter=None, duration_histogram=None, duration=None):
    if is_openai_v1():
        response_dict = model_as_dict(response)
    else:
        response_dict = response

    # metrics record
    _set_chat_metrics(token_counter, choice_counter, duration_histogram, response_dict, duration)

    # span attributes
    _set_response_attributes(span, response_dict)

    if should_send_prompts():
        _set_completions(span, response_dict.get("choices"))

    return response
```

Or maybe even directly inside the `_set_chat_metrics`.

But again, quite some effort for just one URL :)

paolorechia replied:
@nirga what do you think? Should we track this URL as a separate issue?

nirga replied:
@paolorechia not sure I follow - I think this function does return the right base URL, no?

paolorechia replied:
Yes, @Humbertzhang mentioned in a comment, he couldn’t get this URL to work correctly, which why I looked a bit on that

nirga replied:
@Humbertzhang I think this is the logic we should use to get it: https://github.com/traceloop/openllmetry/blob/main/packages/opentelemetry-instrumentation-openai/opentelemetry/instrumentation/openai/shared/__init__.py
(depending on the SDK version, it will be in different attributes).

nirga replied:
For v1, it should be `instance._client.base_url`, and for v0 it should be the code you wrote below

Humbertzhang commented:
> @Humbertzhang thanks for the work here, could you resolve the conflicts so we can merge? :)

Hi @paolorechia, I resolved conflicts in my local env, but when I run tests of traceloop-sdk and openai, I got 
```
FAILED tests/test_privacy_no_prompts.py::test_simple_workflow - openai.APIConnectionError: Connection error.
FAILED tests/test_prompt_management.py::test_prompt_management - openai.APIConnectionError: Connection error.
FAILED tests/test_sdk_initialization.py::test_resource_attributes - openai.APIConnectionError: Connection error.
FAILED tests/test_workflows.py::test_simple_workflow - openai.APIConnectionError: Connection error.
```

and 

```
>           raise CannotOverwriteExistingCassetteException(cassette=cassette, failed_request=vcr_request)
E           vcr.errors.CannotOverwriteExistingCassetteException: Can't overwrite existing cassette ('/Users/maxzhang/Desktop/githubpjs/openllmetry/packages/traceloop-sdk/tests/cassettes/test_workflows/test_simple_workflow.yaml') in your current record mode ('none').
E           No match for the request (<Request (POST) https://api.openai.com/v1/chat/completions>) was found.
E           Found 1 similar requests with 0 different matcher(s) :
E           
E           1 - (<Request (POST) https://api.openai.com/v1/chat/completions>).
E           Matchers succeeded : ['method', 'scheme', 'host', 'port', 'path', 'query']
E           Matchers failed :
```

errors and exceptions.

Have you  or @nirga encountered similar exceptions when using VCR?

nirga commented:
@Humbertzhang I'd try to re-generate the cassettes. Comment out the OpenAI mock key from `conftest.py` and then run `poetry run pytest --record-mode=once`

Humbertzhang commented:
Still got connection errors even i can curl request openai... Maybe u can help me with that @nirga ?
I have pushed my resolve conflict commits.

Humbertzhang commented:
@nirga  not pushed successfully just now😅, now it is pushed

nirga commented:
@Humbertzhang All tests pass now 🤩 
Thanks so much for this, it's a significant project!
Can we just also fix and test the API base as well? (where @paolorechia and I commented above)

Humbertzhang commented:
Hi @nirga  and @paolorechia, I have fixed the `_get_openai_base_url` func (@paolorechia's code and https://github.com/traceloop/openllmetry/pull/522 really helped me, thx!), it now can get url.
I also added assert checks for it in openai's `tests/metrics` tests.

nitin302 commented:
does this also work when used with langchain via from langchain_openai import AzureChatOpenAI?

nirga commented:
@nitin302 no, we need to add support in the langchain instrumentation separately
